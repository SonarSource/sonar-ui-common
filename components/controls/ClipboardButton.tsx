/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import * as Clipboard from 'clipboard';
import Tooltip from './Tooltip';
import { Button } from './buttons';
import { translate } from '../../helpers/l10n';

interface Props {
  className?: string;
  copyValue: string;
  label?: string;
}

interface State {
  tooltipShown: boolean;
}

export default class ClipboardButton extends React.PureComponent<Props, State> {
  clipboard?: Clipboard;
  copyButton?: HTMLElement | null;
  mounted = false;
  state: State = { tooltipShown: false };

  componentDidMount() {
    this.mounted = true;
    if (this.copyButton) {
      this.clipboard = new Clipboard(this.copyButton);
      this.clipboard.on('success', this.showTooltip);
    }
  }

  componentDidUpdate() {
    if (this.clipboard) {
      this.clipboard.destroy();
    }
    if (this.copyButton) {
      this.clipboard = new Clipboard(this.copyButton);
      this.clipboard.on('success', this.showTooltip);
    }
  }

  componentWillUnmount() {
    this.mounted = false;
    if (this.clipboard) {
      this.clipboard.destroy();
    }
  }

  showTooltip = () => {
    if (this.mounted) {
      this.setState({ tooltipShown: true });
      setTimeout(() => {
        if (this.mounted) {
          this.setState({ tooltipShown: false });
        }
      }, 1000);
    }
  };

  render() {
    return (
      <Tooltip overlay={translate('copied_action')} visible={this.state.tooltipShown}>
        <Button
          className={classNames('no-select', this.props.className)}
          data-clipboard-text={this.props.copyValue}
          innerRef={node => (this.copyButton = node)}>
          {this.props.label ? this.props.label : translate('copy')}
        </Button>
      </Tooltip>
    );
  }
}
