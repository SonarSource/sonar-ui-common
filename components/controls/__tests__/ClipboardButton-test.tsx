/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { mount, shallow } from 'enzyme';
import * as React from 'react';
import ClipboardButton from '../ClipboardButton';

const constructor = jest.fn();
const destroy = jest.fn();
const on = jest.fn();

jest.mock(
  'clipboard',
  () =>
    function(...args: any) {
      constructor(...args);
      return {
        destroy,
        on
      };
    }
);

jest.useFakeTimers();

it('should display correctly', () => {
  const wrapper = shallowRender();
  expect(wrapper).toMatchSnapshot();
  wrapper.instance().showTooltip();
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
  jest.runAllTimers();
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});

it('should render a custom label if provided', () => {
  expect(shallowRender({ label: 'Foo Bar' })).toMatchSnapshot();
});

it('should allow its content to be copied', () => {
  const wrapper = mountRender();
  const button = wrapper.find('button').getDOMNode();
  const instance = wrapper.instance();

  expect(constructor).toBeCalledWith(button);
  expect(on).toBeCalledWith('success', instance.showTooltip);

  jest.clearAllMocks();

  wrapper.setProps({ label: 'Some new label' });
  expect(destroy).toBeCalled();
  expect(constructor).toBeCalledWith(button);
  expect(on).toBeCalledWith('success', instance.showTooltip);

  jest.clearAllMocks();

  wrapper.unmount();
  expect(destroy).toBeCalled();
});

function shallowRender(props: Partial<ClipboardButton['props']> = {}) {
  return shallow<ClipboardButton>(createComponent(props));
}

function mountRender(props: Partial<ClipboardButton['props']> = {}) {
  return mount<ClipboardButton>(createComponent(props));
}

function createComponent(props: Partial<ClipboardButton['props']> = {}) {
  return <ClipboardButton copyValue="foo" {...props} />;
}
