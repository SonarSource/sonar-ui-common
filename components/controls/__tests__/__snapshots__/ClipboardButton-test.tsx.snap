// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display correctly 1`] = `
<Tooltip
  overlay="copied_action"
  visible={false}
>
  <Button
    className="no-select"
    data-clipboard-text="foo"
    innerRef={[Function]}
  >
    copy
  </Button>
</Tooltip>
`;

exports[`should display correctly 2`] = `
<Tooltip
  overlay="copied_action"
  visible={true}
>
  <Button
    className="no-select"
    data-clipboard-text="foo"
    innerRef={[Function]}
  >
    copy
  </Button>
</Tooltip>
`;

exports[`should display correctly 3`] = `
<Tooltip
  overlay="copied_action"
  visible={false}
>
  <Button
    className="no-select"
    data-clipboard-text="foo"
    innerRef={[Function]}
  >
    copy
  </Button>
</Tooltip>
`;

exports[`should render a custom label if provided 1`] = `
<Tooltip
  overlay="copied_action"
  visible={false}
>
  <Button
    className="no-select"
    data-clipboard-text="foo"
    innerRef={[Function]}
  >
    Foo Bar
  </Button>
</Tooltip>
`;
